pkg/handler/hooks.go

```diff
@@ -26,23 +26,58 @@ type HookEvent struct {
 	// HTTPRequest contains details about the HTTP request that reached
 	// tusd.
 	HTTPRequest HTTPRequest
+	// Only use by pre-access and pre-create (when Upload-Concat) hook,
+	// for uploads protection for instance
+	Access AccessInfo
 }
 
-func newHookEvent(c *httpContext, info FileInfo) HookEvent {
+type AccessMode = string
+
+const (
+	AccessModeRead  AccessMode = "read"
+	AccessModeWrite AccessMode = "write"
+)
+
+type AccessInfo struct {
+	// read (Head/Get/Upload-Concat) or write (Patch/Delete)
+	Mode AccessMode
+
+	// All files info that will be access by http request
+	// Use an array because of Upload-Concat that may target several files
+	Files []FileInfo
+}
+
+func newHookEvent(c *httpContext, info *FileInfo, accessInfo *AccessInfo) HookEvent {
 	// The Host header field is not present in the header map, see https://pkg.go.dev/net/http#Request:
 	// > For incoming requests, the Host header is promoted to the
 	// > Request.Host field and removed from the Header map.
 	// That's why we add it back manually.
 	c.req.Header.Set("Host", c.req.Host)
 
+	var upload FileInfo
+	if info != nil {
+		upload = *info
+	}
+	var access AccessInfo
+	if accessInfo != nil {
+		access = *accessInfo
+	}
 	return HookEvent{
 		Context: c,
-		Upload:  info,
+		Upload:  upload,
```

Comment:
In hindsight, it would be great if we could set `Upload: nil`, but that would a require a change to the type, which would be breaking.